draft-ietf-core-href.md

```diff
@@ -953,8 +954,180 @@ properties of UTF-8 make this a simple linear process.)
+
+       Otherwise, for each element in the »path« component, include a
+       Uri-Path Option and let that option's value be the text string
+       value of that element.
+
+   8.  If »cri« has a `query` component, then, for each element in the
+       `query` component, include a Uri-Query Option and let that
+       option's value be the be the text string
+       value of that element.
+
+### Composing a Request URI from a Set of CoAP Options {#compose-coap}
+
+   The steps to construct a CRI from a request's options are as follows.
+   These steps either result in a CRI or they fail.
+
+
+[^clarify-host]: Hmm, this is asymmetric.  Is this OK?
+
+
+   1.   Based on the variant of CoAP used in the request, choose a
+        `scheme-id` as per table {{scheme-map}}.  Use that as the first
+        value in the resulting CRI array.
+
+   2.   If the request includes a Uri-Host Option, insert an
+        `authority` with its value as the `host-name`.
+        If that value is not a valid
+        reg-name or IP-literal or IPv4address, fail the algorithm.
+        [^clarify-host]
+
+        If the request does not include a Uri-Host Option, inser an
+        `authority` with `host-ip` being the byte string that
+        represents the request's destination IP address.
+
+   3.   If the request includes a Uri-Port Option, let »port« be that
+        option's value.  Otherwise, let »port« be the request's
+        destination UDP port.
+        If »port« is not the default port for the scheme, then insert
+        the integer value of »port« as the value of `port` in the
+        authority.
+        Otherwise, elide the `port`.
+
+   4.   Insert a `path` component that contains an array built from
+        the text string values of the Uri-Path Options in the request,
+        or an empty array if no such options are present.
+
+   5.   Insert a `query` component that contains an array built from
+        the text string values of the Uri-Query Options in the request,
+        or an empty array if no such options are present.
+
+
+## CoAP Options for Forward-Proxies {#coap-options}
+
+Apart from the above procedures to convert CoAP CRIs to and from sets
+of CoAP Options, two additional CoAP Options are defined in {{Section
+5.10.2 of -coap}} that support requests to forward-proxies:
+
+* Proxy-Uri, and
+* its more lightweight variant, Proxy-Scheme
+
+This section defines analogues of these that employ CRIs and the URI
+Scheme numering provided by the present specification.
+
+### Proxy-CRI
+
+   | No.    | C | U | N | R | Name         | Format | Length | Default |
+   | TBD235 | x | x | - |   | Proxy-Cri    | opaque | 1-1023 | (none)  |
+{: #tab-proxy-cri title="Proxy-Cri CoAP Option"}
+
+The Proxy-CRI Option carries an encoded CBOR data item that represents
+an absolute CRI reference.
+It is used analogously to Proxy-Uri as defined in {{Section 5.10.2
+of -coap}}.
+The Proxy-Cri Option MUST take precedence over any of the Uri-Host,
```

Comment:
Or go full-base, and allow also that there's only the scheme present (at which point Proxy-Scheme-Number becomes somewhat redundant).

Reply:
Examples:

* Request in CoAP-over-TCP to a CoAP resource on a same-host proxcy as described in transport-indication:
  - GET Uri-Path: "foo", "bar", Proxy-Scheme: "coap" (the existing approach)
  - GET Uri-Path: "foo", "bar", Proxy-Scheme-Numeric: 0 (0 byte)
  - GET Uri-Path: "foo", "bar', Cri-SchemeAndAuthority: <<-1>> (1 byte)
  - GET Proxy-Cri is not an option b/c it can't implicitly state the authority.

* Request through a proxy (any transport going forward to UDP literal):
  - GET Uri-Path: "foo", "bar", Proxy-Scheme: "coap", Uri-Host: "[2001:db8::1]" (a hex encoding implementers may dislike)
  - GET Uri-Path: "foo", "bar", Proxy-Scheme-Numeric: 0 (0 byte), Uri-Host: "[2001:db8::1]" (still same issue)
  - GET Uri-Path: "foo", "bar', Cri-SchemeAndAuthority: <<-1, h'20010db8...0001'>>
  - GET Proxy-Cri: <<-1, h'20010db8...0001', ["foo", "bar"]>>

* Request through a proxy (any transport going forward to host name):
  - GET Uri-Path: "foo", "bar", Proxy-Scheme: "coap", Uri-Host: "example.com" (which is a bit string-y but not as bad as the IP literal)
  - GET Uri-Path: "foo", "bar", Proxy-Scheme-Numeric: 0 (0 byte), Uri-Host: "example.com" (still same issue)
  - GET Uri-Path: "foo", "bar', Cri-SchemeAndAuthority: <<-1, "example", "com">>
  - GET Proxy-Cri: <<-1, "example", "com", ["foo", "bar"]>>

* OSCORE splitting of the Proxy-... options, when hitting Proxy-Cri <<0, "example", "com", ["foo", "bar"]>>
  (which this document will probably need to describe to follow the recommendation of OSCORE)
  - could be described as going through conversion to Uri, then split as per 8613, and possibly converted back
  - could be described as splitting the Proxy-Cri into Proxy-Scheme-Numeric: 0, Uri-Host: "example.com" (needs stringification of any IP literal, which we don't have in this case), and inner Uri-Path options
  - could be described as splitting the Proxy-Cri into Cri-SchemeAndAuthority <<-1, "example", "com">> and inner Uri-Path options (provided they even fit in Uri-Scheme, don't know what 8613 says about the equivalent case)